sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReader.scala

```diff
@@ -36,8 +36,15 @@ class FilePartitionReader[T](
 
   private def ignoreMissingFiles = options.ignoreMissingFiles
   private def ignoreCorruptFiles = options.ignoreCorruptFiles
+  private def ignoreCorruptFilesAfterRetries = options.ignoreCorruptFilesAfterRetries
 
   override def next(): Boolean = {
+
+    def shouldSkipCorruptFiles(): Boolean = {
```

Comment:
If network is unreachable or the file storage is broken down in a period of time temporarily, a task which has been retried beyond TASK_MAX_FAILURES would cause data lost.

Reply:
yes, but compare to without this PR, this case is exposed to users instead of silently marking tasks as success.

Reply:
Are you effectively trying to get it to fail and retry on an IOE, rather than not retry because it says, well, I'm ignoring errors anyway? OK. I'm just wondering how many cases that hurts too, where you retry an unrecoverable error longer.

Reply:
In PR desc, I have shown a case where the same query reads the same copy of data at the same time, one is 'corrupt' and the other succeeds. It means that some errors we meet with IOE are not positive data/file corruption. If we ignore them, IMHO, it is a correctness issue. And the config ignoreCorruptFiles says that it ignores corrupt files, not errors. If the data to read contains plenty of corrupt files which means the error is unrecoverable, I add RETIES conf which separates from `spark.task.maxFailures` to limit.

BTW, I have checked the ORC read path, it seems to wrap all errors with IOE.

Reply:
> [DOCS] Whether to ignore corrupt files. If true, the Spark jobs will continue to run when encountering corrupted files and the contents that have been read will still be returned. This configuration is effective only when using file-based sources such as Parquet, JSON and ORC.

FYI, the doc says that it ignores corrupt files, so we are willing to ignore corrupt files, right? For transient failures like a network issue, we want to retry it, because the file is not corrupt. Anyway, like what you said, we can not distinguish whether a failure is transient or permanent, so in this PR, I retry them both with a limit.

> Or I could spin a different argument: if the data can't be read 2 times, and is read a 3rd time, are you OK with that being correct?

For this case, in this PR, we can set ingoreCorruptFilesAfterRetries greater than the maximum task failures, and let it `fail` after retrying enough times. Or maybe we shall rename it to `determineFileCorruptAfterRetries`